river/Output.zig

```diff
@@ -185,227 +163,259 @@ pub fn sendViewTags(self: Self) void {
-    layoutMasterStack(self, visible_count, output_tags, MasterPosition.Top);
-}
 
-/// Wrapper for default layout with master area on the right
-pub fn layoutRightMaster(self: *Self, visible_count: u32, output_tags: u32) void {
-    layoutMasterStack(self, visible_count, output_tags, MasterPosition.Right);
-}
+    if (i != 4) {
+        Log.Error.log("Invalid layout: Expected four window values, got {}.", .{i});
+        return false;
+    }
 
-/// Wrapper for default layout with master area on the bottom
-pub fn layoutBottomMaster(self: *Self, visible_count: u32, output_tags: u32) void {
-    layoutMasterStack(self, visible_count, output_tags, MasterPosition.Bottom);
+    Log.Debug.log("Parsed window layout: x={} y={} width={} height={}", .{ box.x, box.y, box.width, box.height });
+    return true;
 }
 
-/// Wrapper for default layout with master area on the left
-pub fn layoutLeftMaster(self: *Self, visible_count: u32, output_tags: u32) void {
-    layoutMasterStack(self, visible_count, output_tags, MasterPosition.Left);
+test "parse window configuration" {
+    const testing = @import("std").testing;
+    var box: Box = .{ .x = 0, .y = 0, .width = 0, .height = 0 };
+    testing.expect(parseWindowConfig("5 10 100 200", &box));
+    testing.expect(box.x == 5);
+    testing.expect(box.y == 10);
+    testing.expect(box.width == 100);
+    testing.expect(box.height == 200);
 }
 
-/// A layout in which every window uses the maximum available space.
-pub fn layoutFull(self: *Self, visible_count: u32, output_tags: u32) void {
+/// Execute an external layout function, parse its output and apply the layout
+/// to the output.
+fn layoutExternal(self: *Self, visible_count: u32, output_tags: u32) bool {
     const border_width = self.root.server.config.border_width;
     const view_padding = self.root.server.config.view_padding;
     const outer_padding = self.root.server.config.outer_padding;
+    const xy_offset = @intCast(i32, border_width + outer_padding + view_padding);
+    const delta_size = (border_width + view_padding) * 2;
+    const layout_width = @intCast(u32, self.usable_box.width) - outer_padding * 2;
+    const layout_height = @intCast(u32, self.usable_box.height) - outer_padding * 2;
 
-    const layout_width = @intCast(u32, self.usable_box.width) -
-        (outer_padding * 2) - (border_width * 2) - (view_padding * 2);
-    const layout_height = @intCast(u32, self.usable_box.height) -
-        (outer_padding * 2) - (border_width * 2) - (view_padding * 2);
-    const x_offset = self.usable_box.x + @intCast(i32, outer_padding + border_width + view_padding);
-    const y_offset = self.usable_box.y + @intCast(i32, outer_padding + border_width + view_padding);
+    const visible_count_str = std.fmt.allocPrint(self.root.server.allocator, "{}", .{visible_count}) catch |err| switch (err) {
+        error.OutOfMemory => {
+            @panic("out of memory");
+        },
+    };
```

Comment:
These could all be "stringified" at once with one call to `allocPrint`. Something like
```zig
const params = std.fmt.allocPrint(allocator, "{} {} {d} {} {}", .{self.master_count, self.master_factor, ...
```
Also, there's only one possible error here so we can just `catch @panic("out of memory");`